domino/routes.py

```diff
@@ -1,22 +1,35 @@
+from .exceptions import MissingRequiredFieldException
+
 class _Routes:
-    def __init__(self, host, owner_username, project_name):
+    def __init__(self, host, owner_username=None, project_name=None):
         self.host = host
         self._owner_username = owner_username
         self._project_name = project_name
 
     # URL builders
     def _build_project_url(self):
-        return self.host + '/v1/projects/' + \
-            self._owner_username + '/' + self._project_name
+        if self._project_name:
+            return self.host + '/v1/projects/' + \
+                self._owner_username + '/' + self._project_name
+        else:
+            raise MissingRequiredFieldException("Object must be initialised with Project name to use this method")
 
     def _build_project_url_private_api(self):
-        return self.host + '/u/' + self._owner_username + \
-            '/' + self._project_name
+        if self._project_name:
+            return self.host + '/u/' + self._owner_username + \
+                '/' + self._project_name
+        else:
+            raise MissingRequiredFieldException("Object must be initialised with Project name to use this method")
```

Comment:
I would like to see

``` else:
            raise MissingRequiredFieldException("Object must be initialised with Project name to use this method")```

only once and not duplicated into other methods.

Reply:
And no, these checks have to be added to publicly exposed methods requiring projects (I suggested using annotations, like @project_required), not where you place them. This way anybody looking at the code would immediately know if a project is required. And right now many other methods will fail with weird errors, like `job_stop` method from `domino.py`. It requires a project, but your introduced checks won't be called if I'm not mistaken.

Reply:
Look here to better understand what I meant:

https://github.com/dominodatalab/python-domino/pull/104/commits/99d1752b5630c2660f3bb55824f229be24c45fa2

(that solution did not make it to the final PR but a similar one applicable here in my opinion)